framework/src/timesteppers/FunctionDT.C

```diff
@@ -36,23 +41,63 @@ validParams<FunctionDT>()
 
 FunctionDT::FunctionDT(const InputParameters & parameters)
   : TimeStepper(parameters),
+    FunctionInterface(this),
     _time_t(getParam<std::vector<Real>>("time_t")),
     _time_dt(getParam<std::vector<Real>>("time_dt")),
+    _function(nullptr),
     _growth_factor(getParam<Real>("growth_factor")),
     _cutback_occurred(false),
     _min_dt(getParam<Real>("min_dt")),
     _interpolate(getParam<bool>("interpolate"))
 {
-  try
+  // TODO: remove this when `time_t` and `time_dt` is removed
+  if ((isParamValid("time_t") && isParamValid("time_dt")) && !isParamValid("function"))
+    mooseDeprecated(name(),
+                    ": Using `time_t` and `time_dt` parameter is deprecated. Switch your input "
+                    "file to using `function` parameter.\n",
+                    "  1. Build a new function. If `interpolate` parameter is true use type = "
+                    "PiecewiseLinear. If it was false, use PiecewiseConstant.\n",
+                    "  2. Copy `time_t` parameter into your function and rename it to `x`.\n",
+                    "  3. Copy `time_dt` parameter into your function and rename it to `y`.\n",
+                    "  4. Use the `function` parameter in your time stepper and pass your new "
+                    "function name into it.\n");
+  else if ((isParamValid("time_t") && isParamValid("time_dt")) && isParamValid("function"))
+    mooseError(name(),
+               ": Using `time_t`, `_time_dt` and `function` at the same time. Use only `function`, "
+               "`time_t` and _time_dt is deprecated.");
+  else if (!isParamValid("function"))
+    mooseError(name(),
+               ": Please, specify a function (using the `function` parameter) that will prescribe "
+               "the time step size.");
+
+  if (isParamValid("time_t") && isParamValid("time_dt"))
   {
-    _time_ipol = libmesh_make_unique<LinearInterpolation>(_time_t, _time_dt);
+    try
+    {
+      _time_ipol = libmesh_make_unique<LinearInterpolation>(_time_t, _time_dt);
+    }
+    catch (std::domain_error & e)
+    {
+      mooseError("In FunctionDT ", _name, ": ", e.what());
+    }
+
+    _time_knots = _time_t;
+    _use_function = false;
   }
-  catch (std::domain_error & e)
+  else
   {
-    mooseError("In FunctionDT ", _name, ": ", e.what());
+    _function = &getFunction("function");
+    // If dt is given by piece-wise linear and constant function, we add the domain into
+    // _time_knots, so that the time stepper hits those time points
+    if (dynamic_cast<PiecewiseLinear *>(_function) || dynamic_cast<PiecewiseConstant *>(_function))
+    {
+      Piecewise * pw = dynamic_cast<Piecewise *>(_function);
```

Comment:
Why not just cast to Piecewise in the first place (in the "if" statement)?

Reply:
Because there is also `PiecewiseBilinear` and `PiecewiseTrilinear` so this allows 1D functions only. Then, the common `Piecewise` class is very convenient to use